qa/L0_perf_analyzer_capi/test.sh

```diff
@@ -0,0 +1,396 @@
+    echo -e "\n***\n*** Test Failed\n***"
+    exit 1
+fi
+
+export CUDA_VISIBLE_DEVICES=0
+
+CLIENT_LOG="./perf_analyzer.log"
+PERF_ANALYZER=../clients/perf_analyzer
+
+DATADIR=`pwd`/models
+TESTDATADIR=`pwd`/test_data
+
+SERVER_LIBRARY_PATH=/opt/tritonserver
+
+INT_JSONDATAFILE=`pwd`/json_input_data_files/int_data.json
+INT_DIFFSHAPE_JSONDATAFILE=`pwd`/json_input_data_files/int_data_diff_shape.json
+FLOAT_DIFFSHAPE_JSONDATAFILE=`pwd`/json_input_data_files/float_data_with_shape.json
+STRING_JSONDATAFILE=`pwd`/json_input_data_files/string_data.json
+STRING_WITHSHAPE_JSONDATAFILE=`pwd`/json_input_data_files/string_data_with_shape.json
+SEQ_JSONDATAFILE=`pwd`/json_input_data_files/seq_data.json
+SHAPETENSORADTAFILE=`pwd`/json_input_data_files/shape_tensor_data.json
+IMAGE_JSONDATAFILE=`pwd`/json_input_data_files/image_data.json
+
+
+ERROR_STRING="error | Request count: 0 | : 0 infer/sec"
+
+source ../common/util.sh
+
+rm -f $CLIENT_LOG
+rm -rf $DATADIR $TESTDATADIR $ENSEMBLE_DATADIR
+
+mkdir -p $DATADIR
+# Copy fixed-shape models
+cp -r /data/inferenceserver/${REPO_VERSION}/qa_model_repository/graphdef_int32_int32_int32 $DATADIR/
+cp -r /data/inferenceserver/${REPO_VERSION}/qa_model_repository/graphdef_nobatch_int32_int32_int32 $DATADIR/
+cp -r /data/inferenceserver/${REPO_VERSION}/qa_model_repository/graphdef_object_object_object $DATADIR/
+cp -r /data/inferenceserver/${REPO_VERSION}/qa_model_repository/graphdef_nobatch_object_object_object $DATADIR/
+
+# Copy a variable-shape models
+cp -r /data/inferenceserver/${REPO_VERSION}/qa_variable_model_repository/graphdef_object_int32_int32 $DATADIR/
+cp -r /data/inferenceserver/${REPO_VERSION}/qa_variable_model_repository/graphdef_int32_int32_float32 $DATADIR/
+
+# Copy shape tensor models
+cp -r /data/inferenceserver/${REPO_VERSION}/qa_shapetensor_model_repository/plan_zero_1_float32 $DATADIR/
+
+# Copying ensemble including a sequential model
+cp -r /data/inferenceserver/${REPO_VERSION}/qa_sequence_model_repository/savedmodel_sequence_object $DATADIR
+cp -r /data/inferenceserver/${REPO_VERSION}/qa_ensemble_model_repository/qa_sequence_model_repository/simple_savedmodel_sequence_object $DATADIR
+cp -r /data/inferenceserver/${REPO_VERSION}/qa_ensemble_model_repository/qa_sequence_model_repository/nop_TYPE_FP32_-1 $DATADIR
+
+# Copying variable sequence model
+cp -r /data/inferenceserver/${REPO_VERSION}/qa_variable_sequence_model_repository/graphdef_sequence_float32 $DATADIR
+
+mkdir $DATADIR/nop_TYPE_FP32_-1/1
+
+# Copy inception model to the model repository
+cp -r /data/inferenceserver/${REPO_VERSION}/tf_model_store/inception_v1_graphdef $DATADIR
+
+# Copy resnet50v1.5_fp16
+cp -r /data/inferenceserver/${REPO_VERSION}/perf_model_store/resnet50v1.5_fp16_savedmodel $DATADIR
+
+# Generating test data
+mkdir -p $TESTDATADIR
+for INPUT in INPUT0 INPUT1; do
+    for i in {1..16}; do
+        echo '1' >> $TESTDATADIR/${INPUT}
+    done
+done
+RET=0
+########## Test C API #############
```

Comment:
Most of these tests are just testing the operation of different options with perf_analyzer. I don't think it is a good idea to completely replicate the L0_perf_analyzer to work with L0_perf_analyzer_capi. A better way would be to extract just the cases of interest for c_api client backend. We can be more intelligent in designing these tests for more targeted testing..

Reply:
Since we do a test loop for `grpc` and `http`, I think it is fair if we do the same tests for c_api as well. That being said, I'm also unhappy with how there's a lot of duplicated code to just do testing.

Reply:
Please let me know how to do this better?

Reply:
The reason for looping `grpc` and `http` was because we don't really test our C++ client library in CI and use python client library. Our C++ client testing is limited to examples and perf_analyzer tests. C_API on the other hand is being tested by all our end to end tests.
 It's not such a bad idea to add another loop around C_API tests for its simplicity. But if we are going to have a separate L0_per_analyzer_capi then I wanted to highlight that we are not testing anything new with these cases and can be safely removed:
 1. Different input_data options for perf_analyzer. (json, synthetic and directory)
 2. I don't think large tensor size case will add any benefit either.
 3. Request rate manager. (why is this not supported? I believe it interacts with the same interface with client backend)
 4. Runs with a range of concurrency or request rate values.
 The cases which we must be most interested in:
 1. Basic inference with multiple I/O.
 2. String input
 3. Variable shaped tensors
 4. Shape tensors
 5. Async inference (whenever it is supported)
 6. Sequence Model (whenever it is supported)
 7. Shared Memory (whenever it is supported)

Reply:
> However, since the C API does not yet have the same functionality as the other backend, I think it might be cleaner this way, and will be easier to see what this different feature supports. 

I see merit in this point. If you think it will be easier with follow-on work then I am fine resolving this conversation,